texk/ps2pk/mag.1
.\" e
.TH MAG 1 "1 Feb 2016"
.de EX
.sp
.in +5
.nf
.ft CW
..
.de EE
.in -5
.ft R
.sp
.fi
..
.SH NAME
mag \- computes TeX fontsizes and magsteps
.SH SYNOPSIS 
.B
mag
[-\fBR\fIdpi\fP\fP] magstep . . .

.B
mag
[-\fBr\fP] [-\fBR\fP\fIdpi\fP] fontsize . . .

.SH DESCRIPTION
.EQ
delim $$
.EN
This tool calculates fontsizes given magsteps.  TeXfonts are provided as
true sized fonts or as magnifications.  The fontsize of a true sized
font equals the resolution of the printer (ex.  300).  The fontsize
of a font magnified $n$ \fImagsteps\fP equals:
.EQ
1.2 sup{n} times 300
delim off
.EN
rounded to its nearest integer value.  Fontnames for TeX fonts normally
consists of the name (\fIcmr\fP), pointsize (\fI10\fP), type (\fIpk\fP)
and fontsize (\fI360\fP), for example: \fIcmr10.360pk\fP. 

.SH EXAMPLES
The result of \fImag -R240 -2 0 0.5 1\fP will be: 
.EX 
167 240 263 288 
.EE

The inverse function is computed with the \fI-r\fP option. The result of
\fImag -r -R240 167 240 263 288\fP
will be the fontsteps:
.EX
-2 0 0.5 1
.EE

The UNIX shells allow command substitution. One can write:
.EX
mag -r -R240 `mag -R240 -2 0 0.5 1`
.EE

.SH DATE
18 December, 1989

.SH AUTHOR
.nf
Piet Tutelaers
University of Technology Eindhoven
Maintained in TeX Live; please send bug reports or other correspondence
to tex-k@tug.org (http://lists.tug.org/tex-k).


ChangeLog
2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2015 => 2016/dev.
	* tardate.ac: Switch to 2015-06-12.

2015-05-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris (removed): Moved from here to doc/.
	* Makefile.am: Adapted.

2015-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris: New file from Mojca (without very long lines).
	* Makefile.am: Add the new file.

2015-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2015-03-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2015-03-07.

2015-02-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/dist_hook.am (new): Makefile fragment for dist-hook target.
	* Makefile.am: Use the fragment am/dist_hook.am.

2014-11-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-11-24.

2014-11-17  Karl Berry  <karl@tug.org>

	* am/script_links.am: Use w32_wrapper or w64_wrapper.

2014-10-21  Karl Berry  <karl@tug.org>

	* configure.ac: report CC/CXX values, with --version output
	if available.
	* Build: no need for objc any more.

2014-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-08-22.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2014 => 2015/dev.
	* tardate.ac: Switch to 2014-06-16.

2014-04-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Makefile.am: New names for README.* files.

2014-04-17  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2014-04-04  Karl Berry  <karl@tug.org>

	* Build (TL_MAKE,MAKE): set both in all cases.
	(buildenv.log): include pwd, and use .log instead of .txt.

2014-04-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-04-01.

2014-01-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-01-01.

2013-12-26  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (ACLOCAL): Ensure aclocal.m4 is up to date.

2013-10-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* doc: New directory for texlive.info & Co.
	* Makefile.am, configure.ac: Adapted.

2013-10-05  Karl Berry  <karl@tug.org>

	* Build: reorder variables and settings for clarity.
	(TL_CONF_BANNER): restore variable, default is empty string.
	(TL_POSTCONFIGURE): new variable, default true.

2013-09-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Drop ACLOCAL_AMFLAGS.
	* configure.ac: Use AC_CONFIG_MACRO_DIRS.
	Use KPSE_RECURSE_LIBS in m4/kpse-setup.m4 to recurse into all
	TeX specific libraries.

2013-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-09-01.

	* Makefile.am: Moved definition of cf_silent* from here ...
	* am/recurse.am: ... to here 

2013-08-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Add the make target 'texlinks'.

2013-07-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid still more 'parallel make' problems.

2013-07-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid more 'parallel make' problems.

2013-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/*.am: Moved Makefile fragments from libs/am/ and tekk/am/.
	* am/rebuild.am: New Makefile fragment to rebuild a library.
	* am/reconfig.am: New Makefile fragment to reconfigure.
	* am/recurse.am: New Makefile fragment to recurse into subdirs.

	* Makefile.am, configure.ac: Use am/recurse.am.

2013-06-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Added Automake options 'tar-pax' for paths
	longer than 100 chars and 'no-dist-gzip'.
	* version.ac: Switch 2013 => 2014/dev'
	* tardate.ac: Switch to 2013-06-24.

2013-03-26  Karl Berry  <karl@tug.org>

	* Build: use $MAKE rather than $TL_MAKE for the make clean
	in the unlikely event of a local Makefile.

2013-03-21  Mojca Miklavec  <mojca.miklavec.lists@gmail.com>

	* Build: Remove GNUMAKE variable (after freetype2 as the only library
	which needed it has been fixed).

2013-03-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-03-01.

2013-02-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Building for Windows does not require 'ln -s'.
	* README.config: Document '--without-ln-s'.

2013-02-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (mandir, infodir): Change texmf => texmf-dist.

2013-01-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-01-24.
	* Makefile.am: Running texlinks in texk/texlive/tl_scripts.

2013-01-18  Karl Berry  <karl@tug.org>

	* Build: remove use of time.

2012-12-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Enable silent rules.

2012-11-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-11-16.

2012-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-10-24.

2012-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-08-19.

2012-08-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Running texlinks in texk/texlive/linked_scripts.

2012-07-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build: Add TL_COMPILER_GFLAGS also to OBJCXXFLAGS.

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2012 => 2013/dev.

2012-06-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.

2012-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test for GNU make before creating Makefile,
	required for FreeBSD make and perhaps others.

2012-05-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README: Update the list of prerequisites.
	* README.config: Mention configure options for biber.

2012-04-12  Karl Berry  <karl@tug.org>

	* Build (TL_COMPILER_GFLAGS): new envvar, used to override -g
	for all three compilers, with Build -g.

2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>


texk/ps2pk/mag.c
/*
 *	.\" e
 *	.TH MAG 1 UMFT
 *	.SH NAME
 *	mag \- computes fontsizes and magsteps
 *	
 *	.de Ex
 *	.sp
 *	.RS
 *	.nf
 *	.ft C
 *	..
 *	.de Xe
 *	.RE
 *	.sp
 *	.fi
 *	..
 *	.SH SYNOPSIS 
 *	.B
 *	mag
 *	[-\fBR\fIdpi\fP\fP] magstep . . .
 *	
 *	.B
 *	mag
 *	[-\fBr\fP] [-\fBR\fP\fIdpi\fP] fontsize . . .
 *	
 *	.SH DESCRIPTION
 *	.EQ
 *	delim $$
 *	.EN
 *	This tool calculates fontsizes given magsteps.  TeXfonts are provided as
 *	true sized fonts or as magnifications.  The fontsize of a true sized
 *	font equals the resolution of the printer (ex.  300).  The fontsize
 *	of a font magnified $n$ \fImagsteps\fP equals:
 *	.EQ
 *	1.2 sup{n} times 300
 *	delim off
 *	.EN
 *	rounded to its nearest integer value.  Fontnames for TeX fonts normally
 *	consists of the name (\fIcmr\fP), pointsize (\fI10\fP), type (\fIpk\fP)
 *	and fontsize (\fI360\fP), for example: \fIcmr10.360pk\fP. 
 *	
 *	.SH EXAMPLES
 *	The result of \fImag -R240 -2 0 0.5 1\fP will be: 
 *	.Ex 
 *	167 240 263 288 
 *	.Xe
 *	
 *	The inverse function is computed with the \fI-r\fP option. The result of
 *	\fImag -r -R240 167 240 263 288\fP
 *	will be the fontsteps:
 *	.Ex
 *	-2 0 0.5 1
 *	.Xe
 *	
 *	The UNIX shells allow command substitution. One can write:
 *	.Ex
 *	mag -r -R240 `mag -R240 -2 0 0.5 1`
 *	.Xe
 *	
 *	.SH DATE
 *	August, 1992
 *	
 *	.SH AUTHOR
 *	.nf
 *	Piet Tutelaers
 *	University of Technology Eindhoven
 *	rcpt@urc.tue.nl
 */

#include "basics.h"	/* fatal() */

#include <ctype.h> 
#include <math.h>
#include <stdarg.h> 
#include <stdio.h> 
#include <stdlib.h> 
#include <string.h>
#ifdef WIN32
#include <fcntl.h>
#endif

int invert = 0;
float DPI = 300.0;

/* Prototypes */
static int fontsize(double);
static double stepsize(double);

int main(int argc, char *argv[]) {
   float  sz, arg; int c;
   const char *myname = "mag";
   short done;

   while (--argc > 0 && (*++argv)[0] == '-') {
      if (isdigit((int)(*argv)[1])) { /* allow negative numbers as arguments */
         break;
      }
      done=0;
      while ((!done) && (c = *++argv[0]))  /* allow -bcK like options */
      	 switch (c) {
      	 case 'r':
      	    invert = 1; break;
      	 case 'R':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
	    DPI = atof(argv[0]);
	    done = 1;
      	    break;
      	 default:
      	    fatal("%s: %c illegal option\n", myname, c);
      	 }
      }

   if (argc < 1)
      fatal("Usage: %s [-r] [-Rdpi] size . . .\n", myname);
   
#ifdef WIN32
   setmode(fileno(stdout), _O_BINARY);
#endif

   for ( ; argc; argc--, argv++) {
      arg=atof(argv[0]);
      switch (invert) {
      case 0:
         printf("%d%c", fontsize(arg), argc > 1 ? ' ' : '\n');
         break;
      case 1:
         sz=stepsize(arg);
         if (((int)(10*sz))%10==0)
            printf("%d%c", (int)sz, argc > 1 ? ' ' : '\n');
         else 
            printf("%f%c", sz, argc > 1 ? ' ' : '\n');
      }
   }
   return 0;
}

static int
fontsize(double x)
{
   return(DPI*pow(1.2, x) + 0.5);
}

static double
stepsize(double x)
{
   double s;
   s=(log(x)-log(DPI))/log(1.2);
   if (s>=0) return floor(10*s+0.5)/10;
   return -floor(10*(-s)+0.5)/10;
}
